store/rootmulti/simple_map.go

```diff
@@ -0,0 +1,96 @@
+package rootmulti
+
+import (
+	"bytes"
+
+	amino "github.com/tendermint/go-amino"
+	"github.com/tendermint/tendermint/crypto/merkle"
+
+	"github.com/tendermint/tendermint/crypto/tmhash"
+	"github.com/tendermint/tendermint/libs/kv"
+)
+
+// Merkle tree from a map.
+// Leaves are `hash(key) | hash(value)`.
+// Leaves are sorted before Merkle hashing.
+type simpleMap struct {
+	kvs    kv.Pairs
+	sorted bool
+}
+
+func newSimpleMap() *simpleMap {
+	return &simpleMap{
+		kvs:    nil,
+		sorted: false,
+	}
+}
+
+// Set creates a kv pair of the key and the hash of the value,
+// and then appends it to simpleMap's kv pairs.
+func (sm *simpleMap) Set(key string, value []byte) {
+	sm.sorted = false
+
+	// The value is hashed, so you can
+	// check for equality with a cached value (say)
+	// and make a determination to fetch or not.
+	vhash := tmhash.Sum(value)
+
+	sm.kvs = append(sm.kvs, kv.Pair{
+		Key:   []byte(key),
+		Value: vhash,
+	})
+}
+
+// Hash Merkle root hash of items sorted by key
+// (UNSTABLE: and by value too if duplicate key).
+func (sm *simpleMap) Hash() []byte {
+	sm.Sort()
+	return hashKVPairs(sm.kvs)
+}
+
+func (sm *simpleMap) Sort() {
+	if sm.sorted {
+		return
+	}
+	sm.kvs.Sort()
+	sm.sorted = true
+}
+
+// Returns a copy of sorted KVPairs.
+// NOTE these contain the hashed key and value.
+func (sm *simpleMap) KVPairs() kv.Pairs {
+	sm.Sort()
+	kvs := make(kv.Pairs, len(sm.kvs))
+	copy(kvs, sm.kvs)
+	return kvs
+}
+
+//----------------------------------------
+
+// A local extension to KVPair that can be hashed.
+// Key and value are length prefixed and concatenated,
+// then hashed.
+type KVPair kv.Pair
+
+// Bytes returns key || value, with both the
+// key and value length prefixed.
+func (kv KVPair) Bytes() []byte {
+	var b bytes.Buffer
```

Comment:
Why do we have to use amino here? We can easily do this as `len(key) | key | len(value) | value`. Correct me if I'm wrong.

Reply:
/cc @erikgrinaker

Reply:
this type could be removed as well as its not used in the sdk...

Reply:
I think this is just a lift-and-shift from Tendermint, but correct me if I'm wrong. There may well be better ways of doing this, the original issue had some suggestions: https://github.com/tendermint/tendermint/issues/2593.

Reply:
you are correct, i just moved it over. was planning on moving the issue when the pr gets merged. I can make it `len(key) | key | len(value) | value` if people would like it

Reply:
yes, I know it was lift-and-shift. I'm proposing lift-and-shift whiling improving it as you do it. Improve and tidy up the godocs + remove the dep of amino.